### ISSUE TITLE
with*** methods with Option use reference equality checking and not value

### ISSUE THREAD

**Issue by talios** [immutables/immutables#502]
I was recently just adding Google's ErrorProne into our builds and hit the following with the generated Immutables classes using 2.3.9:

  http://errorprone.info/bugpattern/OptionalEquality

    error: [OptionalEquality] Comparison using reference equality instead of value equality
        if (this.password == value) return this;
                          ^
        (see http://errorprone.info/bugpattern/OptionalEquality)
      Did you mean 'if (Objects.equals(this.password, value)) return this;' or 'if (this.password.equals(value)) return this;'?

**Comment by elucash** [comment]
Back then I did this on purpose. The idea is that immutables are preventing to copy values by **shallowly** checking the attributes. In the case of optional it just fine enough to check the wrapper itself. As I don't want to incur full `Object.equals` by design, I'll probably try to unwrap an optional and check underlying value for reference equality, but it will be slightly less performant.

**Comment by elucash** [comment]
Only now, I guess, I've figured it out: the code is using `Optional` in a context where Immutables do not regard it as optional, otherwise we already handle it.

```java
  @SuppressWarning("immutables")
  @Value.Default
  public Optional<Path> getStubBinaryPath() {
    return Optional.empty();
  }
```
@davido .It's best to avoid such constructs ^^ , just use `public abstract Optional<Path> getStubBinaryPath();`. If it's was made to workaround some other problem, please, let me know if we can fix it.

**Comment by talios** [comment]
In my case, we were just using `public abstract Optional<Path> password();` and were getting those warnings.

**Comment by elucash** [comment]
for the default case, not complicated by other conditions, it should be already fixed (our "fixture" module is compiled with ErrorProne, so there is some confidence that it works). Anyway, I'm going to release the new version shortly, which fixes other such cases.

**Comment by davido** [comment]
Ah, yes, I'm seeing that `AbstractApplePlatform.java` from Buck code base conains exactlly this construct for stubBinaryPath, but not for swiftName:

```
/**
   * The Swift name for the platform. For example: {@code ios}. If absent,
   * use {@link #getName()} instead.
   */
  public abstract Optional<String> getSwiftName();

[...]

```
  @SuppressWarnings("immutables")
  @Value.Default
  public Optional<Path> getStubBinaryPath() {
    return Optional.empty();
  }
```

I'm not familiar with this part of Buck code, I just wanted to activate Error Pone on their code base, and running into this issue. I will open an issue on their tracker.

Thanks.

**Comment by elucash** [comment]
released in 2.4.0, thank you!
